src/IO/S3Common.cpp

```diff
@@ -91,28 +107,103 @@ class AWSLogger final : public Aws::Utils::Logging::LogSystemInterface
+            if (diff_seconds > 0 - ACCOUNT_FOR_LATENCY)
+            {
+                auto caller_identity_outcome = sts_client->GetCallerIdentity({});
+                if (caller_identity_outcome.IsSuccess())
+                {
+                    String role_arn = caller_identity_outcome.GetResult().GetArn();
+                    if (role_arn != cached_role_arn || !underlying_credentials_provider)
+                    {
+                        underlying_credentials_provider = std::make_shared<Aws::Auth::STSAssumeRoleCredentialsProvider>(
+                            role_arn, session_name, String(), Aws::Auth::DEFAULT_CREDS_LOAD_FREQ_SECONDS, sts_client
+                        );
+                        cached_role_arn = role_arn;
+                    }
+                    cached_credentials = underlying_credentials_provider->GetAWSCredentials();
+                    expiry = Aws::Utils::DateTime::Now().SecondsWithMSPrecision();
+                }
+                else
+                {
+                    LOG_ERROR(logger, "Failed to make STS GetCallerIdentity request.");
+                }
+            }
+        }
+
+        std::lock_guard<std::mutex> locker(credentials_mutex);
+        return cached_credentials;
+    }
+
+private:
+    std::atomic<int64_t> expiry = 0;
+    std::shared_ptr<Aws::STS::STSClient> sts_client;
+    String session_name;
+    std::mutex credentials_mutex;
+    Aws::Auth::AWSCredentials cached_credentials;
+    std::shared_ptr<Aws::Auth::STSAssumeRoleCredentialsProvider> underlying_credentials_provider;
+    String cached_role_arn;
+};
+
+
 class S3CredentialsProviderChain : public Aws::Auth::AWSCredentialsProviderChain
 {
 public:
-    explicit S3CredentialsProviderChain(const DB::S3::PocoHTTPClientConfiguration & configuration, const Aws::Auth::AWSCredentials & credentials, bool use_environment_credentials)
+    explicit S3CredentialsProviderChain(const DB::S3::PocoHTTPClientConfiguration & configuration, const Aws::Auth::AWSCredentials & credentials, bool use_environment_credentials, bool use_sts_assume_role_credentials)
     {
-        if (use_environment_credentials)
+        auto * logger = &Poco::Logger::get("S3CredentialsProviderChain");
+
+        if (use_sts_assume_role_credentials)
         {
-            const DB::RemoteHostFilter & remote_host_filter = configuration.remote_host_filter;
-            const unsigned int s3_max_redirects = configuration.s3_max_redirects;
+            DB::S3::PocoHTTPClientConfiguration aws_client_configuration = DB::S3::ClientFactory::instance().createClientConfiguration(configuration.remote_host_filter, configuration.s3_max_redirects);
 
+            aws_client_configuration.connectTimeoutMs = 1000;
+            aws_client_configuration.requestTimeoutMs = 1000;
+            aws_client_configuration.retryStrategy = std::make_shared<Aws::Client::DefaultRetryStrategy>(1, 1000);
```

Comment:
Maybe move this constants to config? Timeouts for connect is needed every where in `S3` code, how I think

Reply:
This is not typical read/write but rather getting credentials process. These timeouts are hard coded in AWS SDK.